plutus-tx/src/PlutusTx/DataMap.hs

```diff
@@ -0,0 +1,355 @@
+      BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    go xs =
+      P.matchList
+        xs
+        nil
+        ( \hd tl ->
+            let k' = BI.fst hd
+             in if P.equalsData k k'
+                  then tl
+                  else BI.mkCons hd (go tl)
+        )
+
+{-# INLINEABLE singleton #-}
+singleton :: forall k a. (P.ToData k, P.ToData a) => k -> a -> Map k a
+singleton (P.toBuiltinData -> k) (P.toBuiltinData -> a) = unsafeFromBuiltinList xs
+  where
+    xs = BI.mkCons (BI.mkPairData k a) nil
+
+{-# INLINEABLE empty #-}
+empty :: forall k a. Map k a
+empty = unsafeFromBuiltinList nil
+
+{-# INLINEABLE null #-}
+null :: forall k a. Map k a -> Bool
+null = P.null . toBuiltinList
+
+{-# INLINEABLE unsafeFromList #-}
+unsafeFromList :: (P.ToData k, P.ToData a) => [(k, a)] -> Map k a
+unsafeFromList =
+  unsafeFromBuiltinList
+    . toBuiltin
+    . PlutusTx.Prelude.map (\(k, a) -> (P.toBuiltinData k, P.toBuiltinData a))
+
+{-# INLINEABLE uncons #-}
+uncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  Maybe ((k, a), Map k a)
+uncons m = case P.uncons (toBuiltinList m) of
+  Nothing -> Nothing
+  Just (pair, rest) ->
+    let (k, a) = P.pairToPair pair
+     in Just ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList rest)
+
+{-# INLINEABLE unsafeUncons #-}
+unsafeUncons ::
+  forall k a.
+  (P.UnsafeFromData k, P.UnsafeFromData a) =>
+  Map k a ->
+  ((k, a), Map k a)
+unsafeUncons m =
+  ((P.unsafeFromBuiltinData k, P.unsafeFromBuiltinData a), unsafeFromBuiltinList tl)
+  where
+    (hd, tl) = P.unsafeUncons (toBuiltinList m)
+    (k, a) = P.pairToPair hd
+
+{-# INLINEABLE noDuplicateKeys #-}
+noDuplicateKeys :: forall k a. Map k a -> Bool
+noDuplicateKeys m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let k = BI.fst hd
+             in if member' k tl then False else go tl
+        )
+
+{-# INLINEABLE all #-}
+all :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+all p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        True
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then go tl else False
+        )
+
+{-# INLINEABLE any #-}
+any :: forall k a. (P.UnsafeFromData a) => (a -> Bool) -> Map k a -> Bool
+any p m = go (toBuiltinList m)
+  where
+    go :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData) -> Bool
+    go xs =
+      P.matchList
+        xs
+        False
+        ( \hd tl ->
+            let a = P.unsafeFromBuiltinData (BI.snd hd)
+             in if p a then True else go tl
+        )
+
+{-# INLINEABLE union #-}
+
+-- | Combine two 'Map's.
+union ::
```

Comment:
I'm still failing to gasp what this is doing.  If I have
```
m1 = Map Map [(I 22,I 222),(I 11,I 111)]
m2 = Map Map [(I 22,I 999),(I 33,I 333)]
```
then `union m1 m2` is
```
Map Map [(I 11,Constr 1 [I 111]),(I 22,Constr 3 [I 222,I 999]),(I 33,I 333)]
```
The keys that are in m1 have their values tagged with `This`, the key's that are in both have both values tagged with `These`, but the stuff that's in m2 only is untagged.  Is there a `That` missing is the definition of rs' or somewhere?

Reply:
I was trying to think what would happen if either or both of the inputs contained duplicate keys, but my mind couldn't handle it.